### PR TITLE
Cache Github actions conda env between test runs

### PR DESCRIPTION
**Pull request by LaurenceBeard**
Caches tests between runs by just caching the conda env

For discussion as alternative to #1127

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -17,3 +17,5 @@ dependencies:
   - pycodestyle
   - filelock
   - mock
+  - pip:
+    - clize
```
Comment (TomekTrzeciak) [inline environment.yml]: Nice!

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -2,27 +2,49 @@ name: Tests
...
+        restore-keys: |
+          ${{ runner.OS }}-conda-improver-
+
+    - name: Resolve and install dependencies
+      if: steps.cache.outputs.cache-hit != 'true'
       run: |
-        $CONDA/bin/conda env update -q --file environment.yml --name base
-        $CONDA/bin/pip install clize
-        $CONDA/bin/conda list --export
+        source '/usr/share/miniconda/etc/profile.d/conda.sh'
```
Comment (TomekTrzeciak) [inline .github/workflows/tests.yml]: why is this needed, given that conda is amongst [installed packages on github runners](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/software-installed-on-github-hosted-runners)?

Comment (LaurenceBeard) [reply .github/workflows/tests.yml]: A known problem
https://github.com/conda/conda/issues/7980

Comment (LaurenceBeard) [reply .github/workflows/tests.yml]: writing to .bashrc or similar might be a neater way around this repetition

Comment (TomekTrzeciak) [reply .github/workflows/tests.yml]: Couldn't we just set the `PATH` in the job? That's what `conda activate` boils down to, I think.

https://help.github.com/en/actions/automating-your-workflow-with-github-actions/development-tools-for-github-actions#add-a-system-path-add-path

**Comment by LaurenceBeard** [comment]
Could split these into parallel jobs for pylint/pytest/doctest if we want?

**Comment by TomekTrzeciak** [comment]
> Could split these into parallel jobs for pylint/pytest/doctest if we want?

I wondered about that. I probably wouldn't bother until run-times are above 10-15 min. At the moment all of them put together run well below that, even if you include building conda.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -4,25 +4,29 @@ on: [push, pull_request]
...
+      if: steps.cache.outputs.cache-hit != 'true'
       run: |
-        $CONDA/bin/conda env update -q --file environment.yml --name base
-        $CONDA/bin/pip install clize
-        $CONDA/bin/conda list --export
-
+        source '/usr/share/miniconda/etc/profile.d/conda.sh'
+        conda env update -q --file environment.yml --name improver
+        conda list --export
+    - name: Source conda
```
Comment (TomekTrzeciak) [inline .github/workflows/tests.yml]:
```suggestion
    - name: Activate conda
```
